CMakeLists.txt

```diff
@@ -1,52 +1,78 @@
+set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")
 
-if(NOT SC_BIN_PATH)
-    set(SC_BIN_PATH "${CMAKE_BINARY_DIR}/../bin")
-endif()
-set(SC_EXTENSIONS_DIRECTORY "${SC_BIN_PATH}/extensions")
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${SC_EXTENSIONS_DIRECTORY})
-link_directories(${SC_BIN_PATH} ${SC_EXTENSIONS_DIRECTORY})
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
 
-set(SC_MACHINE_PATH "${PARENT_DIR}/sc-machine")
-set(SC_MEMORY_PATH "${SC_MACHINE_PATH}/sc-memory/sc-memory/include")
-set(SC_CORE_PATH "${SC_MACHINE_PATH}/sc-memory/sc-core/include")
-set(SC_BUILDER_PATH "${SC_MACHINE_PATH}/sc-tools/sc-builder/include")
+if (DEFINED "${SC_EXTENSIONS_DIRECTORY}")
+    message(STATUS "Provided extensions directory: ${SC_EXTENSIONS_DIRECTORY}")
+    file(MAKE_DIRECTORY ${SC_EXTENSIONS_DIRECTORY})
+else()
+    set(SC_EXTENSIONS_DIRECTORY ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/extensions)
+    file(MAKE_DIRECTORY ${SC_EXTENSIONS_DIRECTORY})
+    message(STATUS "No extensions directory provided. Defaulting to: ${SC_EXTENSIONS_DIRECTORY}")
+endif()
 
 option(SCP_CLANG_FORMAT_CODE "Flag to add clangformat and clangformat_check targets" OFF)
-option(SC_BUILD_TESTS "Flag to build unit tests" OFF)
+option(SCP_BUILD_TESTS "Flag to build unit tests" OFF)
 
-if(${SCP_CLANG_FORMAT_CODE})
-    include(${SCP_MACHINE_CMAKE_PATH}/ClangFormat.cmake)
-endif()
+option(AUTO_CCACHE "Use ccache to speed up rebuilds" ON)
 
-if(${SC_BUILD_TESTS})
+if(${SCP_BUILD_TESTS})
     include(CTest)
-    enable_testing()
     include(GoogleTest)
     find_package(GTest REQUIRED)
-    include(${SCP_MACHINE_CMAKE_PATH}/tests.cmake)
+    include(${CMAKE_MODULE_PATH}/tests.cmake)
+endif()
+
+if(${SCP_CLANG_FORMAT_CODE})
+    include(${CMAKE_MODULE_PATH}/ClangFormat.cmake)
 endif()
 
 if(${WIN32})
     message(SEND_ERROR "scp-machine isn't supported on Windows OS.")
 endif(${WIN32})
 
-if(CMAKE_COMPILER_IS_GNUCXX)
-    set(CMAKE_CXX_FLAGS "-std=c++17 -Wall")
-    set(CMAKE_CXX_FLAGS_RELEASE "-O3")        ## Optimize
-    set(CMAKE_EXE_LINKER_FLAGS_RELEASE "-s")  ## Strip binary
+if(${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fcolor-diagnostics")
+elseif(${CMAKE_CXX_COMPILER_ID} MATCHES "GNU")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
 endif()
 
-# for multi-config builds (e.g. msvc)
-foreach(OUTPUT_CONFIG ${CMAKE_CONFIGURATION_TYPES})
-    string(TOUPPER ${OUTPUTCONFIG} OUTPUT_CONFIG)
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${SC_BIN_PATH})
-    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${SC_BIN_PATH})
-endforeach(OUTPUT_CONFIG)
+find_program(CCACHE_PROGRAM ccache)
+if(CCACHE_PROGRAM AND ${AUTO_CCACHE})
+    message(STATUS "Using ${CCACHE_PROGRAM} as compiler launcher")
+    set(CMAKE_CXX_COMPILER_LAUNCHER "${CCACHE_PROGRAM}")
+    set(CMAKE_C_COMPILER_LAUNCHER "${CCACHE_PROGRAM}")  # requires at least CMake 3.9 to be any use
```

Comment:
ok, but 3.24+ is required

Reply:
This comment is to remember that ccache has only been available since this version.